### PR TITLE
Bug 1909600: static-pods: retry failed installers

### PR DESCRIPTION
**Pull request by sttts**
In https://github.com/openshift/api/pull/833 we add `lastFailedTime` and `lastFailedCount` to the static pod node status. This PR adds the corresponding logic:
- when installer fails or operand fails before being ready, we will set the new `lastFailedTime` and increase `lastFailedCount`
- we will use both to compute a retry delay, with exponential backoff, starting with 30s, going up to 10min delay before creating a new installer.
- we choose the next node for upgrade by taking `lastFailedCount` into account, i.e. by preferring such a node with a value != 0 again to continue upgrades.

This will replace the old OOM failure handling, which already today retried the installer pods.

### CODE REVIEW

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 1]
```diff
@@ -857,3 +891,23 @@ func mirrorPodNameForNode(staticPodName, nodeName string) string {
 func statusConfigMapNameForRevision(revision int32) string {
 	return fmt.Sprintf("%s-%d", statusConfigMapName, revision)
 }
+
+func backOffDuration(count int) time.Duration {
+	d := time.Second * time.Duration(float64(10)*math.Pow(1.5, float64(count)))
```
Comment (soltysh) [inline pkg/operator/staticpod/controller/installer/installer_controller.go:896]: Any particular reason not to use client-go's backoff?

**Review by soltysh — COMMENTED** [review]
Few nits, but overall lgtm

File: pkg/operator/staticpod/controller/installer/installer_controller_test.go [hunk 2]
```diff
@@ -1307,18 +1307,19 @@ func TestNodeToStartRevisionWith(t *testing.T) {
 }
 
 func TestSetConditions(t *testing.T) {
-
 	type TestCase struct {
 		name                      string
 		latestAvailableRevision   int32
 		lastFailedRevision        int32
+		lastFailedTime            *metav1.Time
+		lastFailedCount           int
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller_test.go:1451]: I think we decided to have int32 for revisions. This is a normal counter. int is common in the API for that.

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 3]
```diff
@@ -857,3 +891,23 @@ func mirrorPodNameForNode(staticPodName, nodeName string) string {
...
+	if d > time.Minute*10 {
+		return time.Minute * 10
+	}
+	return d
+}
+
+func nthTime(n int) string {
+	switch n % 10 {
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:904]: 11st, 12nd :D

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:904]: but 21st, 22nd.

[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:904]: true, what a language :stuck_out_tongue_closed_eyes:

File: pkg/operator/staticpod/controller/installer/installer_controller.go (hunk 1 above)
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:896]: I looked, but wasn't sure how it fits. This is simple enough to live on its own.

File: pkg/operator/staticpod/controller/installer/installer_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller.go]: fixed

File: pkg/operator/staticpod/controller/installer/installer_controller.go (hunk 1 above)
Comment (soltysh) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:896]: Sure, no objections on that, I was just curious.

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**